src/vmm/src/arch/aarch64/registers.rs

```diff
@@ -0,0 +1,662 @@
+                let reg_data = data.add(self.offset);
+                std::slice::from_raw_parts_mut(reg_data, reg_size)
+            };
+            self.index += 1;
+            self.offset += reg_size;
+            Some(Aarch64RegisterRefMut { id, data })
+        } else {
+            None
+        }
+    }
+}
+
+/// Reference to the aarch64 register.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Aarch64RegisterRef<'a> {
+    /// ID of the register
+    pub id: u64,
+    data: &'a [u8],
+}
+
+impl<'a> Aarch64RegisterRef<'a> {
+    /// Creates new register reference with provided id and data.
+    /// Register size in `id` should be equal to the
+    /// length of the slice. Otherwise this method
+    /// will panic.
+    pub fn new(id: u64, data: &'a [u8]) -> Self {
+        assert_eq!(
+            reg_size(id) as usize,
+            data.len(),
+            "Attempt to create a register reference with incompatible id and data length"
+        );
+
+        Self { id, data }
+    }
+
+    /// Returns register size in bytes
+    pub fn size(&self) -> RegSize {
+        reg_size(self.id).into()
+    }
+
+    /// Returns a register value.
+    /// Type `T` must be of the same length as an
+    /// underlying data slice. Otherwise this method
+    /// will panic.
+    pub fn value<T: Aarch64RegisterData<N>, const N: usize>(&self) -> T {
+        T::from_slice(self.data)
+    }
+
+    /// Returns register data as a byte slice
+    pub fn as_slice(&self) -> &[u8] {
+        self.data
+    }
+}
+
+/// Reference to the aarch64 register.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Aarch64RegisterRefMut<'a> {
+    /// ID of the register
+    pub id: u64,
+    data: &'a mut [u8],
+}
+
+impl<'a> Aarch64RegisterRefMut<'a> {
+    /// Creates new register reference with provided id and data.
+    /// Register size in `id` should be equal to the
+    /// length of the slice. Otherwise this method
+    /// will panic.
+    pub fn new(id: u64, data: &'a mut [u8]) -> Self {
+        assert_eq!(
+            reg_size(id) as usize,
+            data.len(),
+            "Attempt to create a register reference with incompatible id and data length"
+        );
+
+        Self { id, data }
+    }
+
+    /// Returns register size in bytes
+    pub fn size(&self) -> RegSize {
+        reg_size(self.id).into()
+    }
+
+    /// Returns a register value.
+    /// Type `T` must be of the same length as an
+    /// underlying data slice. Otherwise this method
+    /// will panic.
+    pub fn value<T: Aarch64RegisterData<N>, const N: usize>(&self) -> T {
+        T::from_slice(self.data)
+    }
+
+    /// Sets the register value.
+    /// Type `T` must be of the same length as an
+    /// underlying data slice. Otherwise this method
+    /// will panic.
+    pub fn set_value<T: Aarch64RegisterData<N>, const N: usize>(&mut self, value: T) {
+        self.data.copy_from_slice(&value.to_bytes())
+    }
+}
+
+/// Struct describing a aarch64 register.
```

Comment:
I suggest adding a comment advising the type shouldn't be used in new code and explaining why.

Reply:
fixed